### PR TITLE
Upgrader notifications

### LINKED ISSUE #7652: Email notification settings not properly carried forward from 2.0 during upgrade
#### Description
At least 3 settings are not being brought forward correctly:
 - The "receive stuff by email" setting is not brought forward
 - The "follow topics I create & reply to" setting is not brought forward
 - The frequency setting appears to be getting inappropriately changed

**_Question:_**
If you look at the images below, it moves the email requests over as alerts in 2.1, not emails...  Should it move them over as emails?  I believe the assumption is that MOST folks do NOT want emails.  Notifications aren't as intrusive...  

**_E.g., in 2.0:_**
tom-test2-2 0

**_Gets migrated into 2.1 as:_**
tom-test2-2 1

### Environment (complete as necessary)
- Version/Git revision: 2.1.3
- Database Type: MySQL
- Database Version: 8.0
- PHP Version: 8.1

### Additional information/references
Forum discussion:
https://www.simplemachines.org/community/index.php?topic=584199.0

### PR DESCRIPTION
**Pull request by sbulen**
Fixes #7652 

 - Added migration for email announcements setting
 - Added migration for auto-notify setting
 - Fixed migration for regularity setting - different values across 2.0 & 2.1, they are off by 1 

## Test 1
**_2.0_**
U1_before

**_2.1_**
U1_upgr

## Test 2
**_2.0_**
U2_before

**_2.1_**
U2_upgr

## Test 3
**_2.0_**
U3_before

**_2.1_**
U3_upgr

## Test 4
**_
U4_before
2.0_**

**_2.1_*
U4_upgr
*

### CODE REVIEW

File: other/upgrade_2-1_mysql.sql [hunk 1]
```diff
@@ -1060,6 +1061,67 @@ ALTER TABLE {$db_prefix}members
...
+---# Upgrading auto notify setting
+---{
+$_GET['a'] = isset($_GET['a']) ? (int) $_GET['a'] : 0;
+$step_progress['name'] = 'Upgrading auto notify setting';
+$step_progress['current'] = $_GET['a'];
+
+$limit = 100000;
+$is_done = false;
+
+$request = $smcFunc['db_query']('', 'SELECT COUNT(*) FROM {db_prefix}themes WHERE variable = \'auto_notify\'');
```
Comment (albertlast) [inline other/upgrade_2-1_mysql.sql]: Missing the use of string replacement {string:auto_notify} .. array(auto_notify => 'auto_notify')
line break would be also needed in this case.

File: other/upgrade_2-1_postgresql.sql [hunk 2]
```diff
@@ -1261,6 +1262,67 @@ ALTER TABLE {$db_prefix}members
...
+---# Upgrading auto notify setting
+---{
+$_GET['a'] = isset($_GET['a']) ? (int) $_GET['a'] : 0;
+$step_progress['name'] = 'Upgrading auto notify setting';
+$step_progress['current'] = $_GET['a'];
+
+$limit = 100000;
+$is_done = false;
+
+$request = $smcFunc['db_query']('', 'SELECT COUNT(*) FROM {db_prefix}themes WHERE variable = \'auto_notify\'');
```
Comment (albertlast) [inline other/upgrade_2-1_postgresql.sql]: Missing the use of string replacement {string:auto_notify} .. array(auto_notify => 'auto_notify')
line break would be also needed in this case.

File: other/upgrade_2-1_mysql.sql [hunk 3]
```diff
@@ -1060,6 +1061,67 @@ ALTER TABLE {$db_prefix}members
...
+while (!$is_done)
+{
+	nextSubStep($substep);
+	$inserts = array();
+
+	// This setting is stored over in the themes table in 2.0...
+	$request = $smcFunc['db_query']('', '
+		SELECT id_member, value
+		FROM {db_prefix}themes
+		WHERE variable = \'auto_notify\'
```
Comment (albertlast) [inline other/upgrade_2-1_mysql.sql]: Missing the use of string replacement {string:auto_notify} .. array(auto_notify => 'auto_notify')

File: other/upgrade_2-1_postgresql.sql [hunk 4]
```diff
@@ -1261,6 +1262,67 @@ ALTER TABLE {$db_prefix}members
...
+while (!$is_done)
+{
+	nextSubStep($substep);
+	$inserts = array();
+
+	// This setting is stored over in the themes table in 2.0...
+	$request = $smcFunc['db_query']('', '
+		SELECT id_member, value
+		FROM {db_prefix}themes
+		WHERE variable = \'auto_notify\'
```
Comment (albertlast) [inline other/upgrade_2-1_postgresql.sql]: Missing the use of string replacement {string:auto_notify} .. array(auto_notify => 'auto_notify')

File: other/upgrade_2-1_mysql.sql [hunk 5]
```diff
@@ -1060,6 +1061,67 @@ ALTER TABLE {$db_prefix}members
...
+$step_progress['current'] = $_GET['a'];
+
+$limit = 100000;
+$is_done = false;
+
+$request = $smcFunc['db_query']('', 'SELECT COUNT(*) FROM {db_prefix}themes WHERE variable = \'auto_notify\'');
+list($maxMembers) = $smcFunc['db_fetch_row']($request);
+$smcFunc['db_free_result']($request);
+
+while (!$is_done)
```
Comment (albertlast) [inline other/upgrade_2-1_mysql.sql:1084]: while run at least once -> possible issue on multi execute of upgrade runs?

File: other/upgrade_2-1_postgresql.sql [hunk 6]
```diff
@@ -1261,6 +1262,67 @@ ALTER TABLE {$db_prefix}members
...
+$step_progress['current'] = $_GET['a'];
+
+$limit = 100000;
+$is_done = false;
+
+$request = $smcFunc['db_query']('', 'SELECT COUNT(*) FROM {db_prefix}themes WHERE variable = \'auto_notify\'');
+list($maxMembers) = $smcFunc['db_fetch_row']($request);
+$smcFunc['db_free_result']($request);
+
+while (!$is_done)
```
Comment (albertlast) [inline other/upgrade_2-1_postgresql.sql:1285]: while run at least once -> possible issue on multi execute of upgrade runs?

**Review by albertlast — CHANGES_REQUESTED** [review]
biggest concern is, if the logic is fine if they get execute multiple times,
because of the  definition of the while loop.

File: other/upgrade_2-1_postgresql.sql (hunk 6 above)
Comment (jdarwood007) [reply other/upgrade_2-1_postgresql.sql:1285]: I'm not following the issue with multiple runs.  The logic looks sound and we do delete the setting from the table after we are done with this step, which would prevent the issue from multiple runs.

File: other/upgrade_2-1_mysql.sql (hunk 5 above)
Comment (sbulen) [reply other/upgrade_2-1_mysql.sql:1084]: No, should not be an issue due to the insert ignore.

File: other/upgrade_2-1_postgresql.sql (hunk 6 above)
Comment (sbulen) [reply other/upgrade_2-1_postgresql.sql:1285]: Should not be an issue due to the insert ignore.

File: other/upgrade_2-1_mysql.sql (hunk 1 above)
Comment (sbulen) [reply other/upgrade_2-1_mysql.sql]: Done

File: other/upgrade_2-1_mysql.sql (hunk 3 above)
Comment (sbulen) [reply other/upgrade_2-1_mysql.sql]: Done

File: other/upgrade_2-1_postgresql.sql (hunk 4 above)
Comment (sbulen) [reply other/upgrade_2-1_postgresql.sql]: Done

File: other/upgrade_2-1_postgresql.sql (hunk 2 above)
Comment (sbulen) [reply other/upgrade_2-1_postgresql.sql]: Done

**Pull request merged**